### PR TITLE
drivers/flash/soc_flash_nrf: introduce NRF radio scheduler interface

### PR DESCRIPTION
**Pull request by nvlsianpu**
This patch decouples synchronization implementation form NRF flash driver common NVMC related code.
It introduces synchronization interfaces API.

**Rationale of this patch:**
Need to introduces such a enhancement as we foresee different possibilities of mechanism used for synchronization of flash operations with radio operations and other time constrained operations.

**Testing:**
Tested against regression with btshell (using `zephyr/tests/bluetooth/shell`), make sure that `Flash shell` is enabled as well. DUT was nrf52840_pca10056, counterpart  the other nRF52 board.
1. on DUT side:
```console
uart:~$ bt init
uart:~$ bt advertise on
```
2. on counterpart side
```console
uart:~$ bt init
uart:~$ bt scan on
  <printing lot of devices, `test shell` is the one we want to connect>
  [DEVICE]: 5e:de:06:b1:e5:8f (random), AD evt type 4, RSSI -45 test shell
uart:~$ bt scan off
uart:~$ bt connect 5e:de:06:b1:e5:8f (random)
uart:~$ bt conn-update 6 6 0 A // interval 7.5 ms, latency 0, supervisor timeout 100 ms
```
3. Observing whether Link was not lost, on DUT side:
```console
uart:~$ flash erase 0xf9000 0x3000
Erase success.
uart:~$ flash test 0xf9000 0x1000 10
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase-Write test done.

### CODE REVIEW

File: drivers/flash/soc_flash_nrf.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+struct flash_context {
+	uint32_t data_addr;  /* Address of data to write. */
+	uint32_t flash_addr; /* Address of flash to write or erase. */
+	uint32_t len;        /* Size off data to write or erase [B]. */
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+	uint8_t  enable_time_limit; /* set execution limited to the execution
+				     * window.
+				     */
+	void     *synch_context;    /* timing context used by given
```
Comment (hubertmis) [inline drivers/flash/soc_flash_nrf.h:23]: I was thinking if such synchronization backend context could be maintained within synchronization backend module and not exposed to `soc_flash_nrf` at all. It would require changes in `nrfsb_get_write_synch_context()` and `nrfsb_get_erase_synch_context()` declaration to return `void` instead of `void*`; and removal of `void *synch_context` from this structure.

I think it would simplify both sync backends and this module.

File: drivers/flash/soc_flash_nrf.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ */
+int nrfsb_synch_backend_init(void);
+
+/**
+ * Give synchronization context for write operation.
+ *
+ * @retval Synchronization context for the write operation.
+ * Structure of contex depends on backend implementation.
+ */
+void *nrfsb_get_write_synch_context(void);
```
Comment (hubertmis) [inline drivers/flash/soc_flash_nrf.h:72]: I have an idea to remove this function and `nrfsb_get_erase_synch_context()` from API. Instead of preparing write or erase operation, the `soc_flash_nrf` module could pass expected duration of the next flash operation within `struct flash_op_desc` to the synchronization backend. The backend would add its overhead and recalculate timings units according to its requirements.

I understand that it could be less optimal, because synchronization backend would need to calculate operations duration in software for each flash operation. However it looks that API, call flow (number of branch instructions), modules structure would be simplified, what could result in similar performance. But for sure would result in easier maintenance of `soc_flash_nrf` and sync backends.

Any thoughts?

File: drivers/flash/soc_flash_nrf.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * @defgroup nrf_flash_synch_timing synch timing backend API
+ * @ingroup nrf_flash_synch
+ * @{
+ *
+ * API which is used by nrf flash driver for check where execution fill in
+ * the execution window.
+ *
+ * API is used as fallow:
```
Comment (hubertmis) [inline drivers/flash/soc_flash_nrf.h:116]:
```suggestion
 * API is used as follows:
```

File: drivers/flash/soc_flash_nrf.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ * begin of execution window
```
Comment (hubertmis) [inline drivers/flash/soc_flash_nrf.h:117]: Could you add to this example also calls to:
* `nrfsb_get_write_synch_context`
* `nrfsb_work_synchronous_with_radio`
* `flash_op_handler_t handler`
and mention which function is called by which module?

I think it would clarify how this API is used and how to implement it.

File: drivers/flash/soc_flash_nrf.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+/**
+ * Get timestamp and store it in synchronization backend
+ * contex data as opperation beginning time reference.
+ * This timesapm will be used by @ref check_estimated_time_limit()
+ * as the execution window begin referece.
+ *
+ * @param *synch_context timing contex to store the timestamp.
+ *                       Structure of contex depends on backend implementation.
+ */
+void nrfsb_get_timestamp_begin(void *synch_context);
```
Comment (hubertmis) [inline drivers/flash/soc_flash_nrf.h:140]: It seems that this function and `nrfsb_check_estimated_time_limit` could be optional capability of synchronization context. If these functions were not implemented, each chunk of flash operation would be called for consecutive `flash_op_handler_t handler()` calls, what would have worse performance, but implementation would be easier.

We could consider making it optional to implement synchronization backends in steps:
1. PoC without this capability.
2. Final optimal solution with this capability.

I think that for some scheduling algorithms used by synchronization backends, this capability makes no sense and could be omitted even in final solution.

**Review by hubertmis — COMMENTED** [review]
This interface looks nice and implementation looks solid.
I've added some comment with ideas to simplify the interface. Please comment if you think that these ideas are worth implementing, shall we modify them, or discard them.

File: drivers/flash/soc_flash_nrf.h (hunk 1 above)
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h:23]: I agree with you.

File: drivers/flash/soc_flash_nrf.h (hunk 2 above)
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h:72]: I think this is not additional maintenance effort as by the build time configuration timing values (in scientific units) are converted using macros to beck-end implementation own values using macros provide by the ticker infrastructure. What I was thinking introducing per backed defined context was that it is possible that each of them might have original timing variables each other, even different number of orthogonal variables.

@cvinayak What your opinion.

Comment (hubertmis) [reply drivers/flash/soc_flash_nrf.h:72]: Let me rephrase to describe what I meant by increased maintenance.

Synchronization backend does not care if `soc_flash_nrf` performs flash write operation, flash erase, partial erase, partial write, bulk write, bulk erase, or any other operation that I can imagine. It only needs to know duration of the next expected operation. In the proposed interface, the duration is passed by calling either `nrfsb_get_write_synch_context()` or `nrfsb_get_erase_synch_context()`. Now it is responsibility of the backend to deduct expected flash operation timings based on last called function. Each backend would need to contain code that converts and stores expected operation duration based on a call to one or another function.

If the map of flash operation durations would be in `soc_flash_nrf`, it would be in single place. But if we would keep such maps in each backend, similar conversion is copied in multiple modules. Maintaining map in single module is easier than in multiple copies.

One example of increased maintenance would be porting `soc_flash_nrf` driver to a new device with changed flash operations durations. If operations durations were placed only in `soc_flash_nrf.c` and passed as an argument to backends, only 1 file would have to be modified while porting. But if we keep flash operation durations in sync backends, each backend would have to be updated, what is more work and greater possibility of mistake.

Another advantage of keeping duration in `soc_flash_nrf` in single file would be easy extension of the driver. Right now the driver supports only two operations in single product "write" and "erase". When a new use case comes to add "partial erase" operation to the same product that supports typical "erase" (given type of erase could be selected run-time based on number of simultaneous connection), extending `soc_flash_nrf` to add third operation would require changes only in `soc_flash_nrf`. But if we keep current interface, adding new flash operation would require change in `soc_flash_nrf`, alongside the interface and all backend implementations.

Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h:72]: @de-nordic and your opinion?

Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h:72]: I am getting towards the proposal.

Comment (jostn) [reply drivers/flash/soc_flash_nrf.h]: I agree with the proposal from @hubertmis - This will require less maintenance, and gives a more logical abstraction of scheduling back-end vs. flash timings.

Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h]: I am reworking the driver accordingly.

File: drivers/flash/soc_flash_nrf.h [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+ */
+int nrfsb_synch_backend_init(void);
+
+/**
+ * Set synchronization context for synchronous operations.
+ *
+ * This function set beckend's internal context for expected timing parameters.
+ *
+ * @param duration Duration of the execution window [us]
+ * @param interval Interval beatwen execution windows [us]
```
Comment (hubertmis) [inline drivers/flash/soc_flash_nrf.h:90]: Is `interval` required parameter? Is it required that the backend executes flash operations in given interval? What is accepted accuracy of this interval?

I thought that the backend shall execute flash operations as frequently as possible, as soon as possible. Am I mistaken?

Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h:90]: Ticker backed need to know interval (because of ticker/ll design).
But right, in general this could be implementation part to provide this interval.
Will remove interval parameter.

Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h]: Done.
@de-nordic  Actually this change timing while partial erase operation:
previously interval was ~90 ms, and slot was partial erase CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE_MS ms
now interval is the same as slot, so CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE_MS ms

I believe such a long and disproportionate interval was kept by oversight.
I retested the driver for very short partial erase time (3 ms).

File: drivers/flash/soc_flash_nrf.h (hunk 4 above)
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.h:118]: >Could you add to this example also calls to:
* `nrfsb_get_write_synch_context`
* `nrfsb_work_synchronous_with_radio`

Will improve that in the next update

**Review by hubertmis — APPROVED** [review]
The new API looks very good and is clearly documented

**Comment by nvlsianpu** [comment]
I squashed patches.

**Comment by de-nordic** [comment]
Let me be annoying: can we replace `_synch_` and `_synchronize_` everywhere within this code to `_sync_`. Every other code uses `_sync_` . It is easier to grep code if you believe that synchronization stuff contains `_sync_`

**Comment by nvlsianpu** [comment]
>Let me be annoying: can we replace _synch_ and _synchronize_ everywhere within this code to _sync_. Every other code uses _sync_ . It is easier to grep code if you believe that synchronization stuff contains _sync_

done, kept erase/write_synchronously as erase_synch suggest something else than I expected.

File: drivers/flash/soc_flash_nrf_ticker.c [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+	ticks_diff = ticker_ticks_diff_get(ticker_ticks_now_get(),
+					   _ticker_synch_context.ticks_begin);
+	if (ticks_diff + ticks_diff/iteration >
+	    HAL_TICKER_US_TO_TICKS(_ticker_synch_context.slot)) {
+		return true;
+	} else {
+		return false;
+	}
+}
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_ticker.c:224]: Return false by default. Coverity will probably complain about this.

File: drivers/flash/soc_flash_nrf_ticker.c [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+bool nrfsb_is_synch_required(void)
+{
+	return ticker_is_initialized(0);
+}
+
+void nrfsb_get_timestamp_begin(void)
+{
+	_ticker_synch_context.ticks_begin = ticker_ticks_now_get();
+}
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_ticker.c:210]: Could these be given `inline` specifier?

File: drivers/flash/soc_flash_nrf_ticker.c [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+void nrfsb_set_synch_context(uint32_t duration)
+{
+
+	/* FLASH_SYNCH_SWITCHING_TIME is delay which is allways added by
+	 * the slot calling mechanism
+	 */
+	_ticker_synch_context.interval = duration - FLASH_SYNCH_SWITCHING_TIME;
+	_ticker_synch_context.slot = duration;
+}
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf_ticker.c:158]: Maybe give these `inline` specifier, inlining this will take less space than processing call.

File: drivers/flash/soc_flash_nrf.h [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+ * Function executes calbacks op_desc->handler() in execution windows according
+ * to timing settings requested by nrfsb_set_synch_context().
+ * This routine need to be called the handler as many time as it returns
+ * FLASH_OP_ONGOING, howewer an operation timeot should be implemented.
+ * When the handler() returns FLASH_OP_DONE, no further execution windows are
+ * needed so function should return as the handler() finished its operation.
+ *
+ * @retval 0 if op_desc->handler() was executed and
+ * finnished its opperation. Otherwise (timeout, couldn't schedule execution...)
+ * a negative errorcode.
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf.h:100]: I think `errorcode` is missing space.

File: drivers/flash/soc_flash_nrf_ticker.c (hunk 8 above)
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf_ticker.c:210]: ? this is API implementation, can't be inline.

File: drivers/flash/soc_flash_nrf.h [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ *                              execution window
+ *            Driver task           task
+ *                  |                 |
+ *                  |                 |
+ * nrfsb_set_       #                 |
+ * synch_context()  #                 |
+ *                  |                 |
+ *                  |                 |
+ * call nrfsb_      #                 |
```
Comment (de-nordic) [inline drivers/flash/soc_flash_nrf.h:110]: Nice documentation.

File: drivers/flash/soc_flash_nrf_ticker.c (hunk 8 above)
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf_ticker.c:210]: or can be but it don nothing without static...
and can't be static for the other reason...

Comment (de-nordic) [reply drivers/flash/soc_flash_nrf_ticker.c]: I think that it can be `static inline` in header. Have seen this somewhere in Zephyr headers.
Drawback is that if compiler fails to inline function it adds another callable static to object.

**Review by de-nordic — APPROVED** [review]
Not really much to stick to. Looks clear.
Left some comments, but generally approving.

**Comment by nvlsianpu** [comment]
@carlescufi , @ioannisg This is ready to be merged.

File: drivers/flash/soc_flash_nrf.h [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+ * work_sync        #                 |
+ * _with_radio()    #---------------->|
+ *                  |                 |
+ *                  |                 # execution window 0
+ *                  |                 # call flash_op_handler_t handler()
+ *                  |                 #
+ *                  |                 #
+ *                  |                 # flash_op_handler_t handler() return
+ *                  |                 #         FLASH_OP_ONGOING
+ *                  |                 # {beckend request/allow
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:120]: beckend ->  backend

File: drivers/flash/soc_flash_nrf.h [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+ *   check_estimated_time_limit() == 0
+ *   [preserve work context for next execution window]
+ * return form flash_op_handler_t handler()
+ * [return from execution window]
+ * end of execution window
+ */
+
+/**
+ * Get timestamp and store it in synchronization backend
+ * contex data as opperation beginning time reference.
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:173]: contex -> context

File: drivers/flash/soc_flash_nrf.h [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+ * This timesapm will be used by @ref check_estimated_time_limit()
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:174]: timesapm -> timestamp

File: drivers/flash/soc_flash_nrf.h [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+ * as the execution window begin referece.
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:175]: referece -> reference

File: drivers/flash/soc_flash_nrf.h [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+ */
+void nrfsb_get_timestamp_begin(void);
+
+/**
+ * Estimate whether next iteration will fit in time constrains.
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:180]: constrains -> constraints

File: drivers/flash/soc_flash_nrf.h [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+ * This function fetch current timestamp and compare it with the operation
+ * beginning timestam referene stored by @ref get_timestamp_begin() in
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:182]: timestam -> timestamp

File: drivers/flash/soc_flash_nrf.h [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+#include <kernel.h>
+
+#define FLASH_OP_DONE    (0) /* 0 for compliance with the driver API. */
+#define FLASH_OP_ONGOING (-1)
+
+struct flash_context {
+	uint32_t data_addr;  /* Address of data to write. */
+	uint32_t flash_addr; /* Address of flash to write or erase. */
+	uint32_t len;        /* Size off data to write or erase [B]. */
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:18]: off -> of

File: drivers/flash/soc_flash_nrf.h [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+#else
+
+#define FLASH_TIMEOUT_MS ((FLASH_PAGE_ERASE_MAX_TIME_US) * \
+			  (FLASH_PAGE_MAX_CNT) / 1000)
+#endif /* CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE */
+
+/**
+ * @defgroup nrf_flash_sync sync backend API
+ *
+ * API declared bellow contains prototypes of function which shall be
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:46]: bellow -> below

File: drivers/flash/soc_flash_nrf.h [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+ * implemented by the synchronization backend.
+ * @{
+ */
+
+/**
+ * Calback which executes the flash operation.
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:52]: Calback -> Callback

File: drivers/flash/soc_flash_nrf.h [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Check if the operation need to be run synchonous with radio.
+ *
+ * @retval True if operation need to be run synchronously, otherwise False
+ */
+bool nrfsb_is_sync_required(void);
+
+/**
+ * Execute the flash opperation synchronously along the radio operations.
```
Comment (jostn) [inline drivers/flash/soc_flash_nrf.h:89]: opperation -> operation

**Review by jostn — CHANGES_REQUESTED** [review]
Please address spelling mistakes

**Review by jostn — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

File: drivers/flash/soc_flash_nrf.h [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+	flash_op_handler_t handler;
+	struct flash_context *context; /* [in,out] */
+};
+
+/**
+ * Synchronization backend driver initialization procedure.
+ *
+ * This will be run within flash driver initialization
+ */
+int nrfsb_sync_backend_init(void);
```
Comment (carlescufi) [inline drivers/flash/soc_flash_nrf.h:70]: this is not a good prefix. Can we use something like `nrf_flash_sync_*` on all calls instead, and remove the "backend" word?
```
nrf_flash_sync_init()
nrf_flash_sync_set_context()
nrf_flash_sync_is_required();
nrf_flash_sync_exe() / nrf_flash_sync_op()
```

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
@carlescufi done

**Comment by galak** [comment]
@carlescufi can you take a look at this again.

**Review by carlescufi — APPROVED** [review]
